### ISSUE TITLE
Modify proxy handler to preserve context via reflection for property accessors of underlying objects

### ISSUE THREAD

**Issue by mwpastore** [poteto/ember-changeset#429]
I'm not sure if this is a question, bug report, or feature request, so please bear with me...! 

Consider the following example native class and template:

```javascript
import { tracked } from '@glimmer/tracking';

export default class Foo {
  @tracked
  num = 0;

  get doubleNum() {
    return this.num * 2;
  }
}
```

```hbs
{{#let (changeset @myFoo) as |buf|}}
  <input
    type="number"
    value={{changeset-get buf 'num'}}
    {{on 'input' (action (changeset-set buf 'num') value='target.value')}}
  />

  <p>Doubled: {{changeset-get buf 'doubleNum'}}</p>
{{/let}}
```

Regardless of the value of `num` in the changeset—i.e. as the user manipulates the form field—the "Doubled:" value always shows as `0` (zero).  The `this` of `Foo#doubleNum` is always `@myFoo`, not the changeset object. 

I feel like this is a pretty common use case for ember-changeset: we're buffering the changes to `@myFoo`, but we want to show the user what `@myFoo` will look like if and when those changes are applied. In our specific use case, the user is configuring an item in his or her shopping cart, and we want to show the up-to-date unit price of the item as various options are selected and deselected. All while preserving the ability to undo those changes. 

Since we already have access to the "original" value of `doubleNum` based on the unmodified dependent property (e.g. via `@myFoo.doubleNum` or `buf.data.doubleNum`), it seems reasonable to want the proxied value to reflect the modified dependent property. To accomplish this, the `this` of `Foo#doubleNum` needs to be set to the changeset object, or a wrapper. (Easier said than done.)

I'm a little out over my skis on this, but I read [this helpful section](https://javascript.info/proxy#proxying-a-getter) on javascript.info and decided to write my own Changeset factory function that uses a slightly different Proxy handler than the one used by ember-changeset. To wit:

- If the base "key" is present in the changeset, fall back to changeset.get.
- If the "key" or base "key" is a property of the changeset object, fall back to changeset.get. 
- Otherwise, use Reflect.get to retrieve the property of the underlying object, passing the receiver (i.e. the proxy) as the `this' for any property accessors it finds. This functionally supersedes the final two statements of changeset.get. 

TL;DR: It works great! I have some more work to do, and I haven't even started thinking about browser compatibility. I think there's an opportunity to move more (all?) of the logic from changeset.get into the handler but I wanted to start with something manageably small and provable. 

Two major issues:

1. Problem: Since `this` can be a Proxy object, underlying objects can't use `this.constructor` to access static methods and/or properties.
  Workaround: I added `myConstructor = this.constructor;` to the base class of my underlying objects and now we're using `this.myConstructor` for this purpose. But this little "quirk" makes the new proxy handler a breaking change. 
2. Problem: You can't use Reflect.get on tracked properties because the implementation of `@tracked` heavily relies on `this`. The same might be true of computed properties and other things as well. 
  Workaround: Tracked properties (probably?) don't need to be reflected, so I am (very clumsily?) looking for tracked properties and using good ol' Ember.get to fetch them instead.

Here's the code. Feedback more than welcome. Tear it to shreds.

```javascript
import { get as emberGet } from '@ember/object';
import { isPresent } from '@ember/utils';
import { changeset } from 'ember-changeset';

function getPropertyDescriptor(obj, key) {
  if (typeof obj !== 'object' || obj === null || !(key in obj)) {
    return;
  }

  if (Object.prototype.hasOwnProperty.call(obj, key)) {
    return Object.getOwnPropertyDescriptor(obj, key);
  }

  return getPropertyDescriptor(Object.getPrototypeOf(obj), key);
}

function isTracked(obj, key) {
  const descriptor = getPropertyDescriptor(obj, key);

  return descriptor && descriptor.get && /tagForProperty.+consume/s.test(descriptor.get);
}

const proxyHandler = {
  get(target, key, receiver) {
    key = key.toString();

    const [baseKey, remainingPath] = key.split('.', 2);

    if (baseKey in target.change || key in target || baseKey in target) {
      // "default" changeset behavior.
      return target.get(key);
    }

    if (isTracked(target.data, key) || isTracked(target.data, baseKey)) {
      // Can't reflect tracked properties. (And don't need to?)
      return emberGet(target.data, key);
    }

    const result = Reflect.get(target.data, baseKey, receiver);

    // TODO: Use Reflect for nested property accessors as well?
    return isPresent(remainingPath) ? emberGet(result, remainingPath) : result;
  },

  set(target, key, value) {
    target.set(key.toString(), value);

    return true;
  }
};

function Changeset(
  obj,
  validateFn=() => true,
  validationMap={},
  options={}
) {
  const c = changeset(obj, validateFn, validationMap, options);

  return new Proxy(c, proxyHandler);
}

export { Changeset };
```

**Comment by snewcomer** [comment]
> it seems reasonable to want the proxied value to reflect the modified dependent property

100% agree.  I personally haven't caught all the other cases for the Proxy handler.get, we need to clean some things up, and finally, I think this is a great idea!

In order to get this in (I can help as well if you would like), I think a good first step would be an example app with this exact scenario.  What do you think?

**Comment by snewcomer** [comment]
@mwpastore Is this something you wanted me to spike out or were you going to put up a PR?

**Comment by mwpastore** [comment]
@snewcomer I can at least get the ball rolling. Do you still want to start with an example app of this exact scenario, and if so should I just create a new Ember app and make it available on GitHub?

**Comment by snewcomer** [comment]
@mwpastore Perhaps we can take advantage of the dummy app?

**Comment by snewcomer** [comment]
@mwpastore I think it might be best to leave this functionality to the end user (in this case your Proxy handler).  The reason I might suggest this is because I could see other users relying on `doubleNum` to only reflect the property after `save` (applied to underlying model).

I like your implementation.  Does that sound ok to you as long as it work for your app?  Do you have any other thoughts?

**Comment by snewcomer** [comment]
Perhaps try 3.4 to see if anything changed!  Otherwise, I like the idea of a userland Proxy!
